core/test/lib/libledger-test/UvThreadDispatcher.cpp

```diff
@@ -37,190 +37,221 @@
+    //////////////////  UvThreadDispatcher  //////////////////////
+
+    UvThreadDispatcher::UvThreadDispatcher(unsigned int poolSize) : _poolSize(poolSize) {
+        _pool.reserve(_poolSize);
+    }
 
-class ThreadDispatcher : public api::ThreadDispatcher {
-public:
-    std::shared_ptr<api::ExecutionContext> getSerialExecutionContext(const std::string &name) override {
+    std::shared_ptr<ledger::core::api::ExecutionContext> UvThreadDispatcher::getSerialExecutionContext(const std::string &name) {
         std::unique_lock<std::mutex> lock(_mutex);
-        auto it = _serialContexts.find(name);
-        if (it == _serialContexts.end()) {
+        auto it = _contextsByName.find(name);
+        if (it == _contextsByName.end()) {
             auto context = std::make_shared<SequentialExecutionContext>();
-            _serialContexts[name] = context;
+            _pool.push_back(context);
+            _contextsByName[name] = context;
             return context;
         }
         return it->second;
     }
 
-    std::shared_ptr<api::ExecutionContext> getThreadPoolExecutionContext(const std::string &name) override {
-        // TODO Implement thread pools
-        return getSerialExecutionContext(name);
+    std::shared_ptr<ledger::core::api::ExecutionContext> UvThreadDispatcher::getThreadPoolExecutionContext(const std::string &name) {
+        std::unique_lock<std::mutex> lock(_mutex);
+        auto it = _contextsByName.find(name);
+        if (it == _contextsByName.end()) {
+            std::shared_ptr<SequentialExecutionContext> context;
+            if (_pool.size() < _poolSize) {
+                context = std::make_shared<SequentialExecutionContext>();
+                _pool.push_back(context);
+            }
+            else {
+                auto contextIndex = _contextsByName.size() % _poolSize;
+                context = _pool[contextIndex];               
+            }
+            _contextsByName[name] = context;
+            return context;
+        }
+        return it->second;
     }
 
-    std::shared_ptr<api::ExecutionContext> getMainExecutionContext() override {
+    std::shared_ptr<ledger::core::api::ExecutionContext> UvThreadDispatcher::getMainExecutionContext() {
         return getSerialExecutionContext("__uv__main__");
     }
 
-    std::shared_ptr<api::Lock> newLock() override {
+    std::shared_ptr<ledger::core::api::Lock> UvThreadDispatcher::newLock() {
         return nullptr;
     }
 
-    ~ThreadDispatcher() {
-       for (auto& context : _serialContexts) {
-           context.second->stop();
-       }
+    UvThreadDispatcher::~UvThreadDispatcher() {
+        std::unique_lock<std::mutex> lock(_mutex);
+        for (auto& context : _pool) {
+            context->stop();
+            context->waitUntilStopped();
+        }
+    }
+
+    void UvThreadDispatcher::waitUntilStopped() {
+        auto context = std::dynamic_pointer_cast<SequentialExecutionContext>(
+            getSerialExecutionContext("__uv__main__"));
```

Comment:
`__uv__main__` might be a static/class attribute so that no mistake can be done when handling the variable